### PR TITLE
Add precision conversion and quantization filters

### PR DESCRIPTION
**Pull request by ZiyueXu77**
Fixes # .

### Description

Add 16 / 8 / 4 bit quantization filters and apply it to LLM-HF example and experiments

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: examples/advanced/llm_hf/sft_job_compress.py
Comment (holgerroth) [inline examples/advanced/llm_hf/sft_job_compress.py:1]: Why not just add an option for compressions to sft_job?

File: nvflare/app_opt/compression/model_compressor.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            raise ValueError(f"Invalid source data type: {source_data_type}")
+        else:
+            self.source_data_type = source_data_type
+        # check if compression type is valid
+        if compression_type.upper() not in COMPRESSION_TYPE:
+            raise ValueError(f"Invalid compression type: {compression_type}")
+        else:
+            self.compression_type = compression_type
+        # compression constants
+        self.FP16_MIN = np.finfo(np.float32).min
```
Comment (holgerroth) [inline nvflare/app_opt/compression/model_compressor.py:59]: Shouldn't this depend on `source_data_type`?

File: nvflare/app_opt/compression/model_compressor.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+from nvflare.apis.dxo_filter import DXOFilter
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable
+from nvflare.app_opt.compression.constant import COMPRESSION_TYPE, DATA_TYPE
+
+
+class ModelCompressor(DXOFilter):
+    def __init__(
+        self,
+        source_data_type="float32",
```
Comment (holgerroth) [inline nvflare/app_opt/compression/model_compressor.py:31]: in general, why do we need `source_data_type`? Can we not automatically determine it?

File: nvflare/app_opt/compression/model_extractor.py
Comment (holgerroth) [inline nvflare/app_opt/compression/model_extractor.py:1]: I would call this "Decompressor". Extraction is more often used for archives with multiple files. Here, we just use data compression/decompression.

**Review by holgerroth — CHANGES_REQUESTED** [review]
Added some suggestions.

File: examples/advanced/llm_hf/sft_job_compress.py
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/sft_job_compress.py:1]: good point! updated

File: nvflare/app_opt/compression/model_compressor.py (hunk 1 above)
Comment (ZiyueXu77) [reply nvflare/app_opt/compression/model_compressor.py]: yes, initiate here to avoid multiple finfo calling

File: nvflare/app_opt/compression/model_compressor.py (hunk 2 above)
Comment (ZiyueXu77) [reply nvflare/app_opt/compression/model_compressor.py]: removed, also removed its counterpart for decompressor

File: nvflare/app_opt/compression/model_extractor.py
Comment (ZiyueXu77) [reply nvflare/app_opt/compression/model_extractor.py:1]: good idea, changed to "quantizer" and "dequantizer"

File: nvflare/app_opt/quantization/model_dequantizor.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            self.source_data_type = source_data_type
+
+    def dequantization(self, params: dict, quant_state: dict, fl_ctx: FLContext):
+        n_params = len(params.keys())
+        self.log_info(fl_ctx, f"Running dequantization on {n_params} variables")
+        n_bytes_before = 0
+        n_bytes_after = 0
+        n_bytes_meta = 0
+        for i, param_name in enumerate(params.keys()):
+            if self.source_data_type == "float32":
```
Comment (holgerroth) [inline nvflare/app_opt/quantization/model_dequantizor.py:55]: Seems this is the only supported condition. Why check for it?

File: tests/unit_test/app_opt/quantization/quantization_test.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import numpy as np
+import pytest
+
+from nvflare.apis.dxo import DXO, DataKind
+from nvflare.apis.fl_context import FLContext
+from nvflare.app_opt.quantization.model_dequantizor import ModelDequantizor
+from nvflare.app_opt.quantization.model_quantizor import ModelQuantizor
+
+TEST_CASES = [
+    (
```
Comment (holgerroth) [inline tests/unit_test/app_opt/quantization/quantization_test.py:24]: Can you add tests for float16 as well?

File: nvflare/app_opt/quantization/model_dequantizor.py (hunk 3 above)
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_dequantizor.py:55]: for future-proof, potentially we can add more

File: tests/unit_test/app_opt/quantization/quantization_test.py (hunk 4 above)
Comment (ZiyueXu77) [reply tests/unit_test/app_opt/quantization/quantization_test.py:24]: the first test is for float16, and second for block8 (middle line for quant mode)

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Looks good in general. See my comments for thread-safty.

File: nvflare/app_opt/quantization/model_dequantizor.py (hunk 3 above)
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_dequantizor.py:55]: let's check the precision for each sub-vec in case of mixed precision

Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_dequantizor.py:55]: and tensor / other state_dict / numpy

File: tests/unit_test/app_opt/quantization/quantization_test.py (hunk 4 above)
Comment (ZiyueXu77) [reply tests/unit_test/app_opt/quantization/quantization_test.py:24]: check bitsandbytes' offerings (4 bit), provide api for take additional quantizor / dequantizor

File: .github/workflows/premerge.yml [hunk 5]
```diff
@@ -38,6 +38,7 @@ jobs:
         run: |
           python3 -m pip install --upgrade pip
           python3 -m pip install --no-cache-dir -e .[dev]
+          python3 -m pip install --no-cache-dir torch bitsandbytes
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply .github/workflows/premerge.yml:41]: seems to work for this new unit-test (line 24)

File: examples/advanced/llm_hf/README.md [hunk 6]
```diff
@@ -98,6 +98,37 @@ The SFT curves are shown below, black for centralized results, magenta for FL tr
...
+```
+Similarly for SFT, the training loss curves are shown below, black for centralized results, green for FL training with quantization.
+![sft](./figs/fl_sft_comp_8.png)
+
+These results show that model precision conversion / quantization does not significantly impact the training while reducing the message size to 1/2 and 1/4. Thus, it can be important for communicating LLM updates.
+We can also see that the PEFT training loss curves are more aligned than SFT, which is consistent with the results from the centralized training.
+
+For message reduce, for float32 to float16 conversion, the message size is reduced by 2 times. The message size is reduced from 6GB to 3GB for Llama-3.2-1B model according to the log.
+```shell
+Quantized all 147 params. Before quantization: 5993930752 bytes. After quantization: 2996965376 bytes with meta: 0 bytes.
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:125]: sounds good!

File: .github/workflows/premerge.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply .github/workflows/premerge.yml:41]: while this would work, Chester suggest is better, as we use central place for these.

Ziyue could you add that dependency "bitsandbytes" in setup.cfg, after test_support: https://github.com/NVIDIA/NVFlare/blob/main/setup.cfg#L83

[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply .github/workflows/premerge.yml:41]: @ZiyueXu77 actually, I see you have it in app_opt, do you want to put it in APP_OPT's dependency? https://github.com/NVIDIA/NVFlare/blob/main/setup.cfg#L57

File: examples/advanced/llm_hf/README.md [hunk 7]
```diff
@@ -8,15 +8,15 @@ We used the [Llama-3.2-1B model](https://huggingface.co/meta-llama/Llama-3.2-1B)
 
 For PEFT, we used LoRA method, other PEFT methods (e.g. p-tuning, prompt-tuning) can be easily adapted as well by modifying the configs following [PEFT](https://github.com/huggingface/peft) examples.
 
-We would like to showcase two key points in this example:
+We would like to showcase three key points in this example:
 - Adapt local HuggingFace training scripts, both SFT and PEFT, to federated application
 - Handling large model weights (~6 GB for Llama-3.2-1B model with float32 precision for communication), which is beyond protobuf's 2 GB hard limit. It is supported by NVFlare infrastructure via streaming, and does not need any code change.
+- Use NVFlare's filter functionality to enable model quantization and precision conversion for communication, which can significantly reduce the message size and is thus important for communicating LLM updates.
```
Comment (YuanTingHsieh) [inline examples/advanced/llm_hf/README.md:14]: suggest some change:

"which can significantly reduce the message size, making it crucial for transmitting LLM updates."

File: examples/advanced/llm_hf/sft_job.py [hunk 8]
```diff
@@ -139,11 +151,23 @@ def define_parser():
         default="SFT",
         help="training mode, SFT or PEFT, default to SFT",
     )
+    parser.add_argument(
+        "--quantize_mode",
+        type=str,
+        default=None,
+        help="quantization mode, float16 or blockwise8, default to None (no quantization)",
```
Comment (YuanTingHsieh) [inline examples/advanced/llm_hf/sft_job.py:158]: if only two or 3 options, you can do something like:
```
parser.add_argument('--quantize_mode', choices=['train', 'test', 'evaluate'], help='Mode of quantization')
```

File: nvflare/app_opt/quantization/model_dequantizor.py (hunk 3 above)
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_dequantizor.py:55]: check the param type and only perform quantization to numpy float32, will handle tensor for later when we enable transfer tensor directly

File: .github/workflows/premerge.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply .github/workflows/premerge.yml:41]: nice! updated

File: nvflare/app_opt/quantization/model_quantizor.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                    # first clamp the values to the range of float16
+                    values = np.clip(values, self.FP16_MIN, self.FP16_MAX)
+                    # then convert to float16
+                    values = values.astype(np.float16)
+                    n_bytes_after += values.nbytes
+                    params[param_name] = values
+                elif self.quantization_type == "blockwise8":
+                    # use bitsandbytes to quantize the values
+                    # input is a tensor, output is a tuple of (quantized tensor, (absmax, codebook))
+                    # first convert numpy array to tensor
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_quantizor.py:81]: changed the name to NP quantizor / dequantizor, let's handle the tensor case when we enable and properly handle sending tensor directly

File: examples/advanced/llm_hf/README.md (hunk 7 above)
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md:14]: updated

File: examples/advanced/llm_hf/sft_job.py (hunk 8 above)
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/sft_job.py:158]: there can be multiple potentially

File: nvflare/app_opt/quantization/constant.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+# limitations under the License.
+
+DATA_TYPE = [
+    "FLOAT32",
+]
+
+QUANTIZATION_TYPE = [
+    "FLOAT16",
+    "BLOCKWISE8",
+]
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/constant.py:22]: Original data type and the quantization method, so they are separately defined

File: nvflare/app_opt/quantization/model_dequantizor.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        """
+
+        self.log_info(fl_ctx, "Running dequantization...")
+
+        # check config
+        quantization_type = dxo.get_meta_prop(key=MetaKey.PROCESSED_ALGORITHM, default=None)
+        if quantization_type.upper() not in QUANTIZATION_TYPE:
+            raise ValueError(f"Invalid quantization type: {quantization_type}")
+        else:
+            self.quantization_type = quantization_type
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_dequantizor.py:107]: good point! updated.

File: nvflare/app_opt/quantization/model_dequantizor.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+    def dequantization(self, params: dict, quant_state: dict, fl_ctx: FLContext):
+        n_params = len(params.keys())
+        self.log_info(fl_ctx, f"Running dequantization on {n_params} variables")
+        n_bytes_before = 0
+        n_bytes_after = 0
+        n_bytes_meta = 0
+        for i, param_name in enumerate(params.keys()):
+            if self.source_data_type == "float32":
+                if self.quantization_type == "float16":
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_dequantizor.py]: updated with the below

File: nvflare/app_opt/quantization/model_quantizor.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        quant_state = {"absmax": {}, "codebook": {}}
+        for i, param_name in enumerate(params.keys()):
+            values = params[param_name]
+            # check the data type of the values and if it is valid
+            source_data_type = values.dtype.name
+            if source_data_type.upper() not in DATA_TYPE:
+                raise ValueError(f"Invalid source data type: {source_data_type}")
+            # add the number of bytes of the values
+            n_bytes_before += values.nbytes
+            if source_data_type == "float32":
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply nvflare/app_opt/quantization/model_quantizor.py]: updated

**Review by yanchengnv — APPROVED** [review]
LGTM

File: tests/unit_test/app_opt/quantization/quantization_test.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        fl_ctx = FLContext()
+        f_quant = ModelQuantizor(quantization_type=quantization_type)
+        quant_dxo = f_quant.process_dxo(dxo, dxo.to_shareable(), fl_ctx)
+        f_dequant = ModelDequantizor(source_data_type="float32")
+        dequant_dxo = f_dequant.process_dxo(quant_dxo, dxo.to_shareable(), fl_ctx)
+        dequant_data = dequant_dxo.data
+        for key in dequant_data.keys():
+            dequant_array = dequant_data[key]
+            expected_array = expected_data[key]
+            assert np.array_equal(dequant_array, expected_array)
```
[the remark this answers is not in the document]
Comment (holgerroth) [reply tests/unit_test/app_opt/quantization/quantization_test.py:53]: Agreed. This might otherwise fail on different platforms.

File: examples/advanced/llm_hf/README.md [hunk 15]
```diff
@@ -92,12 +92,43 @@ We run the federated training on a single client using NVFlare Simulator via [Jo
...
+python3 sft_job.py --client_ids dolly --data_path ${PWD}/dataset --workspace_dir ${PWD}/workspace/hf_sft_nf4 --job_dir ${PWD}/workspace/jobs/hf_sft_nf4 --train_mode SFT --quantize_mode normfloat4
+```
+The SFT curves are shown below, magenta for centralized results, others for FL training with quantization. We can see it achieves similar alignment comparing to centralized result with training randomness (similar to previous figure).
+![sft](./figs/fl_sft_comp.png)
+
+These results show that model precision conversion / quantization does not significantly impact the training while reducing the message size to 1/2, 1/4, and even 1/8, which can significantly reduce the message size, making it crucial for transmitting LLM updates.
+
+For message reduce, from float32 to 16-/8-/8-bit, the message size is reduced from 6 GB to 3 GB / 1.5 GB / 800 MB for Llama-3.2-1B model according to the log. Note that quantization will generate additional meta data, which can be significant for 4-bit cases.
+
+16-bit:
```
Comment (holgerroth) [inline examples/advanced/llm_hf/README.md:119]: Can you turn this into a table?

File: tests/unit_test/app_opt/quantization/quantization_test.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply tests/unit_test/app_opt/quantization/quantization_test.py:53]: sounds good

File: examples/advanced/llm_hf/README.md (hunk 15 above)
Comment (ZiyueXu77) [reply examples/advanced/llm_hf/README.md]: updated

File: tests/unit_test/app_opt/quantization/quantization_test.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply tests/unit_test/app_opt/quantization/quantization_test.py]: switched to "allclose" rather thabn array_equal

**Review by holgerroth — APPROVED** [review]
Looks good!
